### PR TITLE
Add caching for GatedDeltaNetCache (linear attention models)

### PR DESCRIPTION
**Pull request by Rohan-Bierneni**
### Problem
The Qwen3-Next architecture utilizes a hybrid approach consisting of standard Self-Attention layers and Gated Delta Network (GDN) layers (a form of Linear Attention).
While MaxText’s existing KVCache handles standard attention (growing Key/Value history), it does not support the state management required for GDN. GDN layers require:

  * **A Fixed-size Recurrent State**: Compresses history into a constant-size matrix, rather than growing linearly with sequence length.

  * **A Convolution State**: A sliding window buffer for the short 1D convolution preceding the delta rule.

Without these, Qwen3-Next could not perform efficient autoregressive decoding; it would either need to recompute the entire history at every step or fail to capture temporal dependencies correctly. 

This PR adds caching support to the Gated Delta Net implementation for Qwen3-Next. This is the first non-standard attention caching support in MaxText and required changes to both the MaxEngine code and the Qwen3-Next implementation.

Note: Decoding is currently ONLY supported in unscanned checkpoint format (scan_layers=False and stack_prefill_result_cache=False).

### Solution
This PR implements specific caching infrastructure for Qwen3-Next's linear attention mechanism and integrates it natively into the MaxEngine workflow.

#### 1. New Cache Implementation (src/MaxText/inference/kvcache.py)

  * Introduced GatedDeltaNetCache. Unlike the standard KV cache, this stores fixed-size states:
    * recurrent_state: Shape (B, H, K, V)
    * conv_state: Shape (B, Kernel-1, Dim)
  * Both variables are wrapped in nn.with_logical_constraint to ensure MaxText can automatically infer proper shardings. 
  * Created a BaseCache class to abstract commonalities between standard KV and GDN caches.

#### 2. Layer Logic & State Management (src/MaxText/layers/qwen3.py)

  * Conv & Recurrent Updates: Modified Qwen3NextGatedDeltaNet to properly slide the convolution window and update the recurrent state during MODEL_MODE_AUTOREGRESSIVE.

  * Chunked Prefill Support: Re-wrote MODEL_MODE_PREFILL logic to read from self.cache rather than defaulting to zeros. This allows the GDN state to seamlessly carry over between chunks for long prompts.

  * Dynamic Batch Alignment: Added jnp.broadcast_to and [:batch] slicing during cache retrieval. This dynamically bridges the gap between MaxEngine's global cache allocation (e.g., batch=64) and the active compute shape (e.g., batch=1 during prefill), preventing concatenation shape errors during XLA dummy-tracing passes.

#### 3. MaxEngine Integration (src/MaxText/maxengine.py)

  * Non-sequence-growing Insertions: Updated _insert_jit and bulk_insert to handle fixed-size states. For GDN variables (recurrent_state, conv_state), the prefill step produces a "final state" which is directly copied into the decode cache slot (bypassing the sequence-length slicing logic used for standard KV caches).

  * Packed Prefill Guard: Added a NotImplementedError in insert_partial to block packed prefill for GDN states, preventing sequential memory from bleeding across independent packed prompts.

  * Logical Axes Fix: Updated init_decode_state to properly check for both .logical_axes and .names when unpacking LogicallyPartitioned variables, fixing a bug where empty tuples caused batch index crashes.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/448407748

# Tests

Decode.py tests:
  * On v5p-64 & pdb=1, global_batch_size=32: https://paste.googleplex.com/5916785586077696
  * On CPU vm with pdb=1: https://paste.googleplex.com/5608748317016064

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/decoders.py [hunk 1]
```diff
@@ -895,8 +915,20 @@ def __call__(
...
+                is_full_attention_layer = (lyr + 1) % cfg.inhomogeneous_layer_cycle_interval == 0
+                if is_full_attention_layer:
+                  kv_caches["key_cache"][lyr] = returned_cache[0]
+                  kv_caches["value_cache"][lyr] = returned_cache[1]
+                else:
+                  # gdn_cache = returned_cache.get("gdn_cache")
+                  # if gdn_cache:
+                  #   kv_caches["conv_states"][lyr] = gdn_cache["conv_state"]
+                  #   kv_caches["recurrent_states"][lyr] = gdn_cache["recurrent_state"]
+                  pass
```
Comment (parambole) [inline src/MaxText/layers/decoders.py]: Please remove the commented section.

File: src/MaxText/layers/qwen3.py [hunk 2]
```diff
@@ -326,6 +327,17 @@ def __init__(self, config: Config, dtype: DType = jnp.float32, *, rngs: nnx.Rngs
     conv_kernel_size = cfg.gdn_conv_kernel_dim
     self.v_heads_per_k_head = self.num_v_heads // self.num_k_heads
 
+    if model_mode != MODEL_MODE_TRAIN:
+      self.cache = kvcache.GatedDeltaNetCache(
+          batch=config.per_device_batch_size, # Or appropriate batch dim
```
Comment (parambole) [inline src/MaxText/layers/qwen3.py]: Please remove inline comment here and below if not necessary.

File: src/MaxText/layers/qwen3.py [hunk 3]
```diff
@@ -496,10 +532,31 @@ def __call__(self, hidden_states: Array) -> Array:
...
+        initial_state=recurrent_state,
+        use_qk_norm_in_gdn=cfg.use_qk_norm_in_gdn,
     )
+    
+    if model_mode != MODEL_MODE_TRAIN:
+        # Update self.cache in place for both prefill and decode
+        self.cache.recurrent_state.value = recurrent_state_out
+        
+    # Construct return dictionary for compatibility with decoders.py (optional but safer)
+    new_cache = {
```
Comment (parambole) [inline src/MaxText/layers/qwen3.py]: Is this necessary I don't see it being used?

File: src/MaxText/layers/decoders.py (hunk 1 above)
Comment (Rohan-Bierneni) [reply src/MaxText/layers/decoders.py]: Resolved.

File: src/MaxText/layers/qwen3.py (hunk 2 above)
Comment (Rohan-Bierneni) [reply src/MaxText/layers/qwen3.py]: Resolved.

File: src/MaxText/layers/qwen3.py (hunk 3 above)
Comment (Rohan-Bierneni) [reply src/MaxText/layers/qwen3.py]: Yes, thanks for catching this I removed it and refactored the function signature.

**Review by RissyRan — COMMENTED** [review]
For this one, let's add one unit test like https://github.com/AI-Hypercomputer/maxtext/blob/ec712bf3c7e490bcb0171c6e93e5ac1ac3bdeaa6/tests/unit/attention_test.py#L1195-L1241 to protect any future changes break your feature.

File: tests/unit/attention_test.py [hunk 4]
```diff
@@ -1504,5 +1505,91 @@ def test_tpu_flash_attention_context_parallel(
...
+    for idx in range(prefill_length, decode_total_length):
+      lnx_idx = lnx[:, idx : idx + 1, :]
+
+      gdn_idx = gdn(
+          lnx_idx,
+          model_mode=MODEL_MODE_AUTOREGRESSIVE,
+      )
+
+      gdn_full_this_idx = gdn_full[:, idx : idx + 1, :]
+      self.assertEqual(gdn_full_this_idx.shape, gdn_idx.shape)
```
Comment (RissyRan) [inline tests/unit/attention_test.py:1589]: Is the value matching? The example is from MLA, and I tested locally it's matching. If not, you may want to try dtype=f32.

Comment (Rohan-Bierneni) [reply tests/unit/attention_test.py:1589]: Yes, we are checking the shapes and the values themselves via jax.numpy.allclose(). The test passed locally in TPU vm and should have passed as part of tpu_unit_tests in this PR.

File: src/maxtext/models/qwen3.py [hunk 5]
```diff
@@ -529,15 +545,42 @@ def __call__(self, hidden_states: Array) -> Array:
...
+    conv_state = None
+    if model_mode != MODEL_MODE_TRAIN:
+      # Retrieve state from self.cache
+      conv_state = self.cache.conv_state.value
+
+      if conv_state.shape[0] != batch:
+        if conv_state.shape[0] == 1:
+          conv_state = jnp.broadcast_to(conv_state, (batch,) + conv_state.shape[1:])
+        else:
+          conv_state = conv_state[:batch]
```
Comment (parambole) [inline src/maxtext/models/qwen3.py:565]: It assumes the cache (when shape[0] == 1) is zero-initialized and safe to duplicate.

Just to illustrate the edge case: If a test were to process the word 'Hello' with a batch size of 1, the `conv_state` would update to reflect 'Hello'. If the test then immediately scaled up to a batch size of 2, this logic would broadcast that 'Hello' state into the second slot, meaning both batch items would now share the exact same history.

Could we add a quick inline comment here (e.g., # Assumes zero-initialized state for testing) just to clarify the intent for anyone reading this later?
